src/app/country/country-list/country-list.component.spec.ts

```diff
@@ -0,0 +1,53 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { CountryService } from '../services/country.service';
+import { CountryListComponent } from './country-list.component';
+import { Country } from '../../models/country';
+import { By } from '@angular/platform-browser';
+import { of } from 'rxjs';
+
+jest.mock('../services/country.service');
+
+describe('CountryListComponent', () => {
+  let component: CountryListComponent;
+  let fixture: ComponentFixture<CountryListComponent>;
+  let countryService: CountryService;
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      declarations: [CountryListComponent],
+      providers: [CountryService],
+    }).compileComponents();
+  });
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(CountryListComponent);
+    component = fixture.componentInstance;
+    countryService = TestBed.inject(CountryService);
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should have call to country service', () => {
+    const getCountryServiceSpy = jest.spyOn(countryService, 'getCountries');
+    expect(getCountryServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should select a country and use set contry service', () => {
+    const setCountryServiceSpy = jest.spyOn(countryService, 'setCountry');
+    component.countryList$ = of([
```

Comment:
This is wrong. You're populating ``countryList$`` with the data that is retrieved from your endpoint ``onInit``. So, you need to mock the method that calls the endpoint (and you need to do it before the first fixture.detectChanges(), that initializes the component).

```
beforeEach(() => {
    countryService = TestBed.inject(CountryService);

    jest.spyOn(countryService, 'getCountries').mockReturnValue(of([{ country: 'Afghanistan', city: 'Kabul' }]));

    fixture = TestBed.createComponent(CountryListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });
```

Then your test should be something like:

```
it('should select a country and use set contry service', () => {
    const setCountryServiceSpy = jest.spyOn(countryService, 'setCountry');
    const countryBtns = fixture.debugElement.query(By.css('a.btn')).nativeElement;
    countryBtns.click();
    expect(setCountryServiceSpy).toHaveBeenCalled();
```

Take a look to a similar test in our repo: https://github.com/malaga-front-end/nrwl-angular-jest-cypress-quickstart/blob/master/apps/myapp/src/app/countries/countries.component.spec.ts

Reply:
You are totally right! thanks, I also could update the function because of another comment.